hazelcast/src/main/java/com/hazelcast/internal/util/LatencyDistribution.java

```diff
@@ -0,0 +1,128 @@
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+/**
+ * A latency distribution
+ */
+public final class LatencyDistribution {
+
+    public static final int BUCKET_COUNT = 32;
+
+    public static final String[] LATENCY_KEYS;
+
+    static {
+        LATENCY_KEYS = new String[BUCKET_COUNT];
+        long begin = 0;
+        for (int k = 0; k < BUCKET_COUNT; k++) {
+            long max = LatencyDistribution.bucketMaxUs(k);
+            LATENCY_KEYS[k] = begin + ".." + (max) + "us";
+            begin = max;
+        }
+    }
+
+    private static final AtomicLongFieldUpdater<LatencyDistribution> COUNT
+            = newUpdater(LatencyDistribution.class, "count");
+    private static final AtomicLongFieldUpdater<LatencyDistribution> TOTAL_MICROS
+            = newUpdater(LatencyDistribution.class, "totalMicros");
+    private static final AtomicLongFieldUpdater<LatencyDistribution> MAX_MICROS
+            = newUpdater(LatencyDistribution.class, "maxMicros");
+
+    private final AtomicLongArray buckets = new AtomicLongArray(BUCKET_COUNT);
+
+    private volatile long count;
+    private volatile long maxMicros;
+    private volatile long totalMicros;
+
+    public int bucketCount() {
+        return BUCKET_COUNT;
+    }
+
+    public long bucket(int bucket) {
+        return buckets.get(bucket);
+    }
+
+    public static long bucketMaxUs(int bucket) {
+        return 1L << bucket;
+    }
+
+    public long count() {
+        return count;
+    }
+
+    public long maxMicros() {
+        return maxMicros;
+    }
+
+    public long totalMicros() {
+        return totalMicros;
+    }
+
+    public long avgMicros() {
+        return count == 0 ? 0 : totalMicros / count;
+    }
+
+    public void done(long startNanos) {
+        recordNanos(System.nanoTime() - startNanos);
+    }
+
+    public void recordNanos(long durationNanos) {
+        // nano clock is not guaranteed to be monotonic. So lets record it as zero so we can at least count.
+        if (durationNanos < 0) {
+            durationNanos = 0;
+        }
+
+        // can lead to problems for calls taking more than 33 days.
+        int durationMicros = (int) NANOSECONDS.toMicros(durationNanos);
+
+        COUNT.addAndGet(this, 1);
+        TOTAL_MICROS.addAndGet(this, durationMicros);
+
+        for (; ; ) {
+            long currentMax = maxMicros;
+            if (durationMicros <= currentMax) {
+                break;
+            }
+
+            if (MAX_MICROS.compareAndSet(this, currentMax, durationMicros)) {
+                break;
+            }
+        }
+
+        try {
+            buckets.incrementAndGet(usToBucketIndex(durationMicros));
+        } catch (RuntimeException e) {
+            throw new RuntimeException("duration nanos:" + durationNanos, e);
+        }
+    }
+
+    public static int usToBucketIndex(int us) {
+        // This logic can be improved. Currently we always round up to the next bucket.
```

Comment:
Why don't we improve the logic now?

Reply:
I was already working on it.

Reply:
The accuracy factor = 0.9991395066958436

That is good enough :)